lib/main.ts

```diff
@@ -0,0 +1,144 @@
+}
+
+/**
+ * Top-level entry point to programs written in Effection. That means that your
+ * program should only call `main` once, and everything the program does is
+ * handled from within `main` including an orderly shutdown. Unlike `run`, `main`
+ * automatically prints errors that occurred to the console.
+ *
+ * Use the {@link exit} operation form within to halt program execution
+ * immediately and initiate shutdown.
+ *
+ * The behavior of `main` is slightly different depending on the environment it
+ * is running in.
+ *
+ * ### Deno, Node
+ *
+ * When running within Deno or Node, any error which reaches `main` causes the
+ * entire process to exit with an exit code of `1`.
+ *
+ * Additionally, handlers for `SIGINT` are attached to the
+ * process, so that sending an exit signal to it causes the main task
+ * to become halted. This means that hitting CTRL-C on an Effection program
+ * using `main` will cause an orderly shutdown and run all cleanup code.
+ *
+ * > Warning! do not call `Deno.exit()` on Deno or `process.exit()` on Node
+ * > directly, as this will not gracefully shutdown. Instead, use the
+ * > {@link exit} operation.
+ *
+ * ### Browser
+ *
+ * When running in a browser, The `main` operation gets shut down on the
+ * `unload` event.
+ */
+
+export async function main(
+  op: (args: string[]) => Operation<void>,
+): Promise<void> {
+  let hardexit = (_status: number) => {};
+
+  let result = await run(() => action<Exit>(function* (resolve) {
+    // action will return shutdown immediately upon resolve, so stash
+    // this function in the exit context so it can be called anywhere.
+    yield* ExitContext.set(resolve);
+
+    try {
+      let interrupt = () => resolve({ status: 130, signal: "SIGINT " });
+      return yield* withHost({
+        *deno() {
+          hardexit = (status) => Deno.exit(status);
+          try {
+            Deno.addSignalListener("SIGINT", interrupt);
+            yield* op(Deno.args.slice());
+          } finally {
+            Deno.removeSignalListener("SIGINT", interrupt);
+          }
+        },
+        *node() {
+          //@ts-expect-error type-checked by Deno, run on Node
+          hardexit = (status) => global.process.exit(status);
+          try {
+            //@ts-expect-error type-checked by Deno, run on Node
+            process.on("SIGINT", interrupt);
+            //@ts-expect-error type-checked by Deno, run on Node
+            yield* op(global.process.args.slice());
+          } finally {
+            //@ts-expect-error this runs on Node
+            process.off("SIGINT", interrupt);
+          }
+        },
+        *browser() {
+          try {
+            self.addEventListener("unload", interrupt);
+            yield* op([]);
+          } finally {
+            self.removeEventListener("unload", interrupt);
+          }
+        },
+      });
+    } catch (error) {
+      resolve({ status: 1, error });
+    }
+
+    yield* exit(0);
+  }));
+
+  if (result.message) {
```

Comment:
What do you think about returning a `Result` type to be internally consistent?

Reply:
It's worth considering as it is very similar to a `Result`, but I'm not sure if the `Result` type fits because it is a value that represents a JavaScript outcome which has a clean boundary between success (`value: T`) and failure(`error: Error`). On the other hand, the type of `result` here needs to represent a _POSIX process outcome_ and the line between success and failure is a bit more blurry in my mind. For example, is a signal exit an error? what value of `Error` would we use to represent that.